### PR TITLE
Add configuration value: compute.shortcut_limit

### LINKED ISSUE #715: Convert global or limit-related variables into configurations.
https://github.com/databricks/koalas/commit/94adf06d622aaac1eb05c03407db8a1abc01632c added configuration base like Pandas. See https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html

Currently we have multiple variables that should be converted into configuration values. We should identify and convert it to configuration _all_.

- [x] Support typed values in configurations (#714)
- [x] Add 'display.max_rows' (#714)
- [x] Add 'compute.shortcut_limit' (#717) https://github.com/databricks/koalas/blob/8360c586f5ad4c8847fe394eefbd77b48d95f687/databricks/koalas/frame.py#L373-L379
- [x] Add 'compute.max_rows'  (#721) https://github.com/databricks/koalas/blob/8360c586f5ad4c8847fe394eefbd77b48d95f687/databricks/koalas/frame.py#L1463-L1470 (I think we should remove `limit` parameter).
- [x] Add 'compute.shortcut_limit' (#717) https://github.com/databricks/koalas/blob/8360c586f5ad4c8847fe394eefbd77b48d95f687/databricks/koalas/groupby.py#L806-L812
- [x] Add 'compute.shortcut_limit' (#717) https://github.com/databricks/koalas/blob/8360c586f5ad4c8847fe394eefbd77b48d95f687/databricks/koalas/groupby.py#L1512-L1518
- [x] Add 'plotting.max_rows'  (#728) https://github.com/databricks/koalas/blob/8360c586f5ad4c8847fe394eefbd77b48d95f687/databricks/koalas/plot.py#L54
- [x] Add 'plotting.sample_ratio' (#737) https://github.com/databricks/koalas/blob/8360c586f5ad4c8847fe394eefbd77b48d95f687/databricks/koalas/plot.py#L85
- [x] Add 'compute.ops_on_diff_frames' (#725) https://github.com/databricks/koalas/blob/8360c586f5ad4c8847fe394eefbd77b48d95f687/databricks/koalas/utils.py#L63
- [x] Add 'compute.default_index_type' (#723) https://github.com/databricks/koalas/blob/8360c586f5ad4c8847fe394eefbd77b48d95f687/databricks/koalas/internal.py#L477
- [x] Add 'compute.max_rows' (#736) https://github.com/databricks/koalas/pull/712/files#diff-5ba644c40e914732959b9b4867f30c53R1644
- [x] Add 'compute.shortcut_limit' (#739) https://github.com/databricks/koalas/pull/732/files#diff-5ba644c40e914732959b9b4867f30c53R1605
- [ ] Add documentation (#743). See https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html

### PR DESCRIPTION
**Pull request by itholic**
related with #715 i added compute.shorcut_limit

Resolves #715

### CODE REVIEW

File: databricks/koalas/config.py [hunk 1]
```diff
@@ -34,6 +34,7 @@
     # For example, this value determines whether the repr() for a dataframe prints out fully or
     # just a truncated repr.
     "display.max_rows": 1000,  # TODO: None should support unlimited.
+    "compute.shortcut_limit": 1000,
```
Comment (HyukjinKwon) [inline databricks/koalas/config.py]: Can you add description for this configuration as above one?

File: databricks/koalas/groupby.py [hunk 2]
```diff
@@ -803,7 +804,7 @@ def apply(self, func):
         if should_infer_schema:
             # Here we execute with the first 1000 to get the return type.
             # If the records were less than 1000, it uses pandas API directly for a shortcut.
-            limit = 1000
+            limit = get_option("compute.shortcut_limit")
```
Comment (HyukjinKwon) [inline databricks/koalas/groupby.py]: Let's fix related tests together - https://github.com/databricks/koalas/blob/master/databricks/koalas/tests/test_groupby.py#L373

File: databricks/koalas/frame.py [hunk 3]
```diff
@@ -373,7 +373,7 @@ def _reduce_for_stat_function(self, sfun, name, axis=None, numeric_only=False):
         elif axis in ('columns', 1):
             # Here we execute with the first 1000 to get the return type.
             # If the records were less than 1000, it uses pandas API directly for a shortcut.
-            limit = 1000
+            limit = get_option("compute.shortcut_limit")
```
Comment (HyukjinKwon) [inline databricks/koalas/frame.py]: this test too https://github.com/databricks/koalas/commit/d354de649ebdea0f75705e2923a1bfb355371ca5#diff-d3f67c2eab2a8dd3df7283426995819fR60

File: databricks/koalas/groupby.py (hunk 2 above)
Comment (itholic) [reply databricks/koalas/groupby.py]: @HyukjinKwon Thnaks, I've fixed them all :)

File: databricks/koalas/config.py [hunk 4]
```diff
@@ -34,6 +34,11 @@
     # For example, this value determines whether the repr() for a dataframe prints out fully or
     # just a truncated repr.
     "display.max_rows": 1000,  # TODO: None should support unlimited.
+
+    # compute.shortcut_limit sets the shortcut value by using collected pandas dataframe directly.
+    # When the input value over this will executed by PySpark.
+    # Set `None` to unlimit.
```
Comment (HyukjinKwon) [inline databricks/koalas/config.py]: Let's remove `Set `None` to unlimit.`

File: databricks/koalas/config.py [hunk 5]
```diff
@@ -34,6 +34,11 @@
     # For example, this value determines whether the repr() for a dataframe prints out fully or
     # just a truncated repr.
     "display.max_rows": 1000,  # TODO: None should support unlimited.
+
+    # compute.shortcut_limit sets the shortcut value by using collected pandas dataframe directly.
+    # When the input value over this will executed by PySpark.
```
Comment (HyukjinKwon) [inline databricks/koalas/config.py]: `When the dataframe length is larger than this limit, Koalas users PySpark to compute`

File: databricks/koalas/tests/test_groupby.py [hunk 6]
```diff
@@ -370,28 +371,34 @@ def test_apply(self):
...
+            })
+            kdf = koalas.DataFrame(pdf)
+
+            self.assert_eq(
+                kdf.groupby('car_id').apply(lambda _: pd.DataFrame({"column": [0.0]})).sort_index(),
+                pdf.groupby('car_id').apply(lambda _: pd.DataFrame({"column": [0.0]})).sort_index())
+        finally:
+            reset_option('compute.shortcut_limit')
 
     def test_transform(self):
```
Comment (HyukjinKwon) [inline databricks/koalas/tests/test_groupby.py]: Can you fix this test too?

Comment (HyukjinKwon) [inline databricks/koalas/tests/test_groupby.py]: And, while we're here, can you remove those tests in `test_filter`:

```
        # Data is intentionally big to test when schema inference is on.
        pdf = pd.DataFrame({'a': [1, 2, 3, 4, 5, 6] * 300,
                            'b': [1, 1, 2, 3, 5, 8] * 300,
                            'c': [1, 4, 9, 16, 25, 36] * 300}, columns=['a', 'b', 'c'])
        kdf = koalas.DataFrame(pdf)
        self.assert_eq(kdf.groupby("b").filter(lambda x: x.b.mean() < 4).sort_index(),
                       pdf.groupby("b").filter(lambda x: x.b.mean() < 4).sort_index())
        self.assert_eq(kdf.groupby(['a', 'b']).filter(lambda x: any(x.a == 2)).sort_index(),
                       pdf.groupby(['a', 'b']).filter(lambda x: any(x.a == 2)).sort_index())
        with self.assertRaisesRegex(TypeError, "<class 'int'> object is not callable"):
            kdf.groupby("b").filter(1)
```

`groupby.filter` actually doesn't infer its schema. It seems mistakenly the test was added.

File: databricks/koalas/config.py [hunk 7]
```diff
@@ -34,6 +34,11 @@
     # For example, this value determines whether the repr() for a dataframe prints out fully or
     # just a truncated repr.
     "display.max_rows": 1000,  # TODO: None should support unlimited.
+
+    # compute.shortcut_limit sets the shortcut value by using collected pandas dataframe directly.
```
Comment (HyukjinKwon) [inline databricks/koalas/config.py]: 'compute.shortcut_limit' sets the limit for a shortcut. It computes specified number of rows and use its schema.

**Comment by HyukjinKwon** [comment]
Looks good otherwise.

**Comment by itholic** [comment]
@HyukjinKwon Thanks, I've resolved all you commented.

File: databricks/koalas/tests/test_groupby.py [hunk 8]
```diff
@@ -344,80 +345,96 @@ def test_bfill(self):
...
+            })
+            kdf = koalas.DataFrame(pdf)
+
+            self.assert_eq(
+                kdf.groupby('car_id').apply(lambda _: pd.DataFrame({"column": [0.0]})).sort_index(),
+                pdf.groupby('car_id').apply(lambda _: pd.DataFrame({"column": [0.0]})).sort_index())
+
+            # 1000+ records will only infer the schema.
+            pdf = pd.DataFrame({
+                "timestamp": [0.0, 0.5, 1.0, 0.0, 0.5],
```
Comment (HyukjinKwon) [inline databricks/koalas/tests/test_groupby.py]: There seems a mistake here. Can you `"timestamp": [0.0, 0.5, 1.0, 0.0, 0.5],` -> `"timestamp": [0.0, 0.5, 1.0, 0.0, 0.5] * 300,` to make it 1000+ records?

Comment (itholic) [reply databricks/koalas/tests/test_groupby.py]: @HyukjinKwon Thanks to comment. i've fixed them :)

**Comment by HyukjinKwon** [comment]
Sorry, it needs to resolve conflicts again.

BTW, if you write `Resolves #715` or `Closes #715` in the PR description, it will resolve and close the issue #715 automatically when this PR is merged

**Comment by itholic** [comment]
@HyukjinKwon Thank you for the good tips! And The conflicts has fixed.

**Comment by HyukjinKwon** [comment]
I added the PR description to show `Resolves #715` usage. Also, let's don't cc somebody in PR description. it should only contain things about what this PR proposes.

**Comment by itholic** [comment]
@HyukjinKwon Thanks to comment! and i think maybe we resolve all `Add` things related with 'compute.shortcut_limit' in  #717 . Could you check them out when you available?

**Pull request merged**